### ISSUE TITLE
Units for experimental Young's modulus data

### ISSUE THREAD

**Issue by eriktamsen** [BAMresearch/LebeDigital#153]
While working on the calibration PR, I had a look at the extracted experimental data for the Young's modulus test.
The units are part of the name of the columns in the CSV file. This is not ideal, or usable.
As I have lost the overview of what is happening in which branch and issue, I am not sure if this has already been adressed.
As far as I can remember, all relevant fields in the ontology have units linked.
I assume therefore the data extraction script needs to be improved, to solve this, so that the units are availiable in a machine readble format. I would suggest to add them to the metadata yaml (or json if that has been changed).
It is possible, that the units dont change as our data extraction is only for one specific machine. If this is the case, this would still require the information to be availiable. Meaning, e.g. just hard coding this data in the meta data file.

@AidaZt,  @ThiloMuth, @StephanPirskawetz: Please check if any change is required and if so, make a plan of how to best implement this.

**Comment by AidaZt** [comment]
There is no placeholder currently in knowledge graph templates to store the information of the units. This ones mentioned in #118 , I think Alina is looking into it. @alFrie right?

**Comment by joergfunger** [comment]
I think this automatic conversion creates more problems than it solves. I would rather have an error message (e.g. in the SHACL validation when certain assumptions (see above) are not fulfilled.

**Comment by joergfunger** [comment]
1. is a valid point, but I think that could be handled afterwards or even in a query (I remember this discussion in the ontology playground, and there seem to be ways to handle that in the query). e.g. [here](https://stackoverflow.com/questions/57560215/querying-distance-units-in-sparql-wikidata-example) and [here](https://en.wikibooks.org/wiki/SPARQL/WIKIDATA_Precision,_Units_and_Coordinates#Normalised_units). As such I would start with just specifying the unit, if we consider a conversion required we could still use e.g. [pint ](https://pint.readthedocs.io/en/stable/) for an automatic conversion in a later stage.
2. Why is that tricky to map the units? I understand that it is not just a string, but a link to an instance defined in  in the PMD core, I guess the units are specified there. Then we just need a conversion function that maps a unit string (e.g. MPa) to the corresponding entity. An idea might be to use another sign (instead of $$ $$), or always specify the expected conversion - usually a string ($$string myvalue$$) or for units ($$unit MPa$$) in the templates, the first one just replacing the string myvalue (would work also for real data, int, since that is essentially a string in the ttl file) and the second one (with unit) calling a function that has as input a string (MPa) and as output the required URI of the MPa unit.

A predefinition is not optimal, because e.g. in the US, everyone is still using kpi (kilopond per square inch). If we force them to use a different unit, it just would not work.

**Comment by mattheokru** [comment]
To 2.
![Screenshot (59)](https://user-images.githubusercontent.com/113978670/234499428-6aed35cf-82cc-42b1-873e-bbf7512833e1.png)

This could be a way to map units. Like Jörg suggested the placeholder just needs to be replaced with the correct unit (an example is in the picture of a pmd unit at the bottom). I could create a Table in which every unit is listed in correspondence to its pmd individual.
Would this work?

**Comment by eriktamsen** [comment]
Whichever way the units will be integrated in the KG, we need to specify them in the extracted metadata.

**Comment by joergfunger** [comment]
I guess ns3:Q174789 is mm in PMD units (are they using qudt, or reinventing PMD units?). Why is the placeholder http://... and not just Diameter_Unit? And as you suggested, there could by a python function (with a table, e.g. a dict) that maps the string in the json file (key, e.g. N/mm2 or MPa) to the ns3:Q17478 (value in the dict). In order to know, that this function has to be used to transform the string, I suggested to either use different placeholder markers, e.g. # or have a "labeled" placeholder (see above). 
Just an update on how this could be integrated into the sparql query, there is a [paper](https://www.mdpi.com/2673-8244/3/1/3) and a [zenodo link](https://zenodo.org/record/7469686#.ZEjvB85BxaQ) with example query that convert e.g. feet to m.

**Comment by mattheokru** [comment]
I think they are reinventing PMD units, however as I looked into the QUDT maybe we should use this Ontology for the units since it has far more units than the current PMD?
The Placeholders for the Units have an http://... because I needed to create a dummy Individual because otherwise ontopanel Converter wouldn't let me export the Ontology with an Individual which doesn't fulfill the requirements for an Individial (full URI)
The Placeholders are currently $$ for the units, like the Values, but I will change them to ## (e.g ##SpecimenDiameter_Unit##)

**Comment by mattheokru** [comment]
In addition the QUDT already has a "dictionary" for the units:
![Screenshot (60)](https://user-images.githubusercontent.com/113978670/234555940-32cf9738-3941-4e85-b4ce-08f5b23caa77.png)

**Comment by joergfunger** [comment]
@mattheokru Why is that different now for the placeholders for the units compared to the other strings. The others also do not have the https ...?

**Comment by StephanPirskawetz** [comment]
Look for GM:
http://qudt.org/vocab/unit/GM
:)

**Comment by joergfunger** [comment]
Yes, it makes sense to give that a unit, since the unit of the raw data file (columns) has to be stored somewhere. Theoretically, they could also be added as metadata to the file, but I think adding it to the graph here makes more sense.

**Comment by joergfunger** [comment]
DateTimeStamp is probably the starting point, and ExperimentTime is the duration of the experiment (the offset). In the optimal setting, this information would be stored in the description of the classes in the ontology, such that questions like that are self-explaining. @StephanPirskawetz  is that included? Both should have units, the first one including a complete date , the second one only a duration). As for the unit of duration, @StephanPirskawetz will have to answer that, but I think s is correct.

**Comment by eriktamsen** [comment]
@alFrie: Ich hoffe ich verstehe deine Frage an Stephan richtig.
Die Daten, die im E-Modul Test der BAM aufgezeichnet wurden, also in den "specimen.dat", haben auch eine Zeile mit Einheiten (Zeile 15). Die können und sollten also auch direkt dort ausgelesen werden für die Metadaten.
Die "Transducers" messen eine Verschiebung, in diesem Fall in mm.
Die Kraft (CompressionForce) ist gemessen in Kilonewton.

**Comment by AidaZt** [comment]
@alFrie you can also check the units in ExampleDataCPTOMapping.xlsx from Mattheo, the one i sent you.

**Comment by mattheokru** [comment]
@alFrie The Experiment Date is the date and time the experiment took place (that's why this is a date time stamp), the Experiment Time is suppose the be the duration of the test (hence the unit seconds).
The Measurement Duration was initially suppose to tell the time the measurement took place (linked the data with the time within the test), but since we linked an excel column for the compression force we should do the same for the measurement duration, however I'm not sure if this is really necessary for the Ontology? If not I would remove MeasurementDuration.
![Screenshot (62)](https://user-images.githubusercontent.com/113978670/236824113-39596a8e-b567-48cd-a321-3cfd2ba48b4d.png)

**Comment by mattheokru** [comment]
The desciption of the Classes/Individuals unfortunatly gets removed if you export them through Ontopanel

**Comment by StephanPirskawetz** [comment]
> Danke @StephanPirskawetz !
> 
> > Die Einheiten in Zeile 15 sind für die Zeitreihen in den nachfolgenden Zeilen und spaltenweise zugeordnet. Da, soweit ich verstanden habe, diese Rohdaten (hier noch) nicht beschrieben werden, können wir alles nach dem Datum ignorieren.
> 
> So hatte ich das auch verstanden. Und wie gesagt gibt es dafür ja auch gar keine Platzhalter. Deshalb würde ich die Daten in der blauen Box / Zeile 15 nicht extrahieren.
> 
> > Nach gelb umrandet: Start des Programmteils Belastungsfunktion. Das ist die wichtige Zeit und die sollte in die Metadaten übernommen werden.
> 
> Also die im neuen Bild grün umrandete Box ist nun die wichtige, ja? Willst du die als eine Variable (also Datum und Uhrzeit gemeinsam) oder getrennt als "ExperimentTime" und "ExperimentDate" wie bisher abgespeichert haben? Wenn alle anderen Zeiten, selbst die gelb umrandete, irrelevant sind, dann sollte die Instanz "MeasurementDuration" ja aus dem KnowledgeGraph gelöscht werden, richtig?
> 
> ![image](https://user-images.githubusercontent.com/93983136/237071691-3211b9c1-612d-433c-8f0b-8879aec9cf74.png)

Ja, grün ist wichtig! Ob gemeinsam oder getrennt weiß ich nicht. Es gibt als Datentypen allerdings nur "dateTime". Wenn man mit dateTime rechnen kann, was ich vermute, dann ist das besser. Wenn auch im Mix KG so ein dateTime vorkommt, kann man die vermutlich einfach voneinander abziehen und bekommt so das Probenalter zur Zeit der Prüfung. Das wäre praktisch. Ist mit getrennten Variablen wahrscheinlich schwieriger.

**Comment by StephanPirskawetz** [comment]
Ja, wir können die Zeit einfach auf 12:00 Uhr setzen. Falls es wichtig ist (Prüfungen früher als nach 28 Tagen), muss die Zeit der Wasserzugabe mit angegeben werden.

**Comment by eriktamsen** [comment]
> > ![image](https://user-images.githubusercontent.com/93983136/237071691-3211b9c1-612d-433c-8f0b-8879aec9cf74.png)
> 
> Ja, grün ist wichtig! Ob gemeinsam oder getrennt weiß ich nicht. Es gibt als Datentypen allerdings nur "dateTime". Wenn man mit dateTime rechnen kann, was ich vermute, dann ist das besser. Wenn auch im Mix KG so ein dateTime vorkommt, kann man die vermutlich einfach voneinander abziehen und bekommt so das Probenalter zur Zeit der Prüfung. Das wäre praktisch. Ist mit getrennten Variablen wahrscheinlich schwieriger.

Also die grüne Box ist für uns auch wichtig.

> Die Einheiten in Zeile 15 sind für die Zeitreihen in den nachfolgenden Zeilen und spaltenweise zugeordnet. Da, soweit ich verstanden habe, diese Rohdaten (hier noch) nicht beschrieben werden, können wir alles nach dem Datum ignorieren.

@StephanPirskawetz: Meiner Meinung nach ist das so nicht richtig. Für unsere Simulation müssen wir wissen mit welchen Einheiten wir arbeiten. Selbst wenn du sagt, dass die BAM Daten alle die gleichen Einheiten haben, dann sollte das aber trotzdem irgendwo maschinenlesbar hinterlegt werden. Warum sollte das nicht einfach teil der Metadaten sein?

**Comment by mattheokru** [comment]
@alFrie Ja sollte da sein. Da das alles etwas chaos war habe ich einen Branch erstellt für die KG's. So ist alles an einem Platz und nicht in verschiedenen Branches/Pull Requests verteilt. Das macht es einfacher den Überblick zu behalten. Der Branch ist hier: https://github.com/BAMresearch/LebeDigital/tree/Knowledge_Graphs_Update
